File: app/src/main/java/com/chess/cryptobot/content/Preferences.java
package com.chess.cryptobot.content;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import java.util.HashSet;
import java.util.Set;

public abstract class Preferences {
    private Context context;
    private SharedPreferences sharedPreferences;
    private String preferenceKey;

    public Preferences(Context context) {
        this.context = context;
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        this.preferenceKey = initPrefKey(context);
    }

    public abstract String initPrefKey(Context context);

    private String getPreferenceKey() {
        return preferenceKey;
    }

    public Set<String> getItems() {
        return new HashSet<>(sharedPreferences.getStringSet(preferenceKey, new HashSet<>()));
    }

    public void setItems(Set<String> items) {
        updateItemSet(items);
    }

    void addItem(String itemName) {
        Set<String> itemsSet = getItems();
        if (itemsSet.contains(itemName)) return;
        itemsSet.add(itemName);
        updateItemSet(itemsSet);
    }

    void removeItem(String itemName) {
        Set<String> itemsSet = getItems();
        itemsSet.remove(itemName);
        updateItemSet(itemsSet);
    }

    private void updateItemSet(Set<String> itemsSet) {
        SharedPreferences.Editor editor = getSharedPreferences().edit();
        editor.putStringSet(getPreferenceKey(), itemsSet);
        editor.apply();
    }

    public SharedPreferences getSharedPreferences() {
        return sharedPreferences;
    }

    public Context getContext() {
        return context;
    }
}

File: app/src/main/java/com/chess/cryptobot/content/Preferences.kt
package com.chess.cryptobot.content

import android.content.Context
import android.content.SharedPreferences
import androidx.preference.PreferenceManager
import androidx.core.content.edit
import kotlin.collections.HashSet

abstract class Preferences protected constructor(context: Context) {
    val sharedPreferences: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
    private val preferenceKey: String by lazy { initPrefKey(context) }

    var items: HashSet<String>?
        get() = sharedPreferences.getStringSet(preferenceKey, HashSet())?.toHashSet()
        set(items) = updateItemSet(items)

    protected abstract fun initPrefKey(context: Context): String

    open fun addItem(itemName: String) {
        val itemsSet = items
        if (itemsSet != null) {
            if (itemsSet.contains(itemName)) return
        }
        itemsSet?.add(itemName)
        updateItemSet(itemsSet)
    }

    internal fun removeItem(itemName: String) {
        val itemsSet = items
        itemsSet?.remove(itemName)
        updateItemSet(itemsSet)
    }

    private fun updateItemSet(itemsSet: HashSet<String>?) {
        sharedPreferences.edit { putStringSet(preferenceKey, itemsSet) }
        val editor = sharedPreferences.edit()
        editor.putStringSet(preferenceKey, itemsSet)
        editor.apply()
    }
}
